backend/common/src/main/java/ai/verta/modeldb/common/futures/InternalFuture.java

```diff
@@ -11,28 +16,52 @@
 import java.util.stream.Collectors;
 
 public class InternalFuture<T> {
+  private static final boolean DEEP_TRACING_ENABLED;
+  public static final AttributeKey<String> STACK_ATTRIBUTE_KEY = AttributeKey.stringKey("stack");
+
+  static {
+    String internalFutureTracingEnabled = System.getenv("IFUTURE_TRACING_ENABLED");
+    if (internalFutureTracingEnabled == null) {
+      DEEP_TRACING_ENABLED = false;
+    } else {
+      DEEP_TRACING_ENABLED = Boolean.parseBoolean(internalFutureTracingEnabled);
+    }
+  }
+
+  private final Tracer futureTracer = GlobalOpenTelemetry.getTracer("futureTracer");
+  private final String formattedStack;
   private CompletionStage<T> stage;
 
   // keep the calling OpenTelemetry context, so we can use it to wrap the actual invocation of the
   // future's implementation
   private final Context callingContext = Context.current();
 
-  private InternalFuture() {}
+  private InternalFuture() {
+    if (!DEEP_TRACING_ENABLED) {
+      formattedStack = null;
+    } else {
+      StackTraceElement[] stackTrace = Thread.currentThread().getStackTrace();
+      if (stackTrace.length > 10) {
+        stackTrace = Arrays.copyOf(stackTrace, 10);
```

Comment:
we'll probably want to refine this in the future, but this will get us info for now